Add support for 3.13

sydney-runkle opened the pull request:
Closes https://github.com/pydantic/pydantic/issues/9668 (add 3.13 support)
Addresses https://github.com/pydantic/pydantic/issues/9613 for the V2 case

sydney-runkle commented:
TODOs here:

- [ ] fix issues in v1 with `_eval_type`
- [ ] fix issues in `eval_type_backport` with `eval_type`
- [ ] fix issues with `Annotated` and `issubclass`

There's probably more...

sydney-runkle commented:
Python 3.13 requires a newer version of `typing-extensions` - was there a reason good reason why we're stuck on 4.6.1 in `pyproject.toml`?

dmontagu commented on pydantic/_internal/_typing_extra.py:
```diff
@@ -109,10 +109,8 @@ def all_literal_values(type_: type[Any]) -> list[Any]:
-
     origin = get_origin(ann_type)
-    return origin is not None and lenient_issubclass(origin, Annotated)
+    return (origin is not None) and (origin is Annotated)
```
Why not just origin is Annotated

sydney-runkle replied:
Done :)

sydney-runkle commented on pydantic/_internal/_typing_extra.py:
```diff
@@ -109,10 +109,8 @@ def all_literal_values(type_: type[Any]) -> list[Any]:
-
     origin = get_origin(ann_type)
-    return origin is not None and lenient_issubclass(origin, Annotated)
+    return (origin is not None) and (origin is Annotated)
```
```suggestion
    return origin is Annotated
```

dmontagu commented on tests/test_computed_fields.py:
```diff
@@ -427,7 +427,10 @@ def _double(self) -> int:
-@pytest.mark.skipif(sys.version_info < (3, 9), reason='@computed_field @classmethod @property only works in 3.9+')
+@pytest.mark.skipif(
+    sys.version_info < (3, 9) or sys.version_info >= (3, 13),
+    reason='@computed_field @classmethod @property only works in 3.9-3.12',
```
Is there a way to do this in 3.13?

sydney-runkle replied:
I didn't find an easy way - I suppose you could create a custom decorator that basically handles the `classmethod` stuff with `__wrapped__` as suggested here:

https://github.com/python/cpython/pull/92379/files#diff-78f24041d66ab8ed2ae1aee94bcd42396d27af833cb96a3c506294c6d6dce82d

alexmojaki replied:
Should a new issue be opened for this?

alexmojaki commented on pydantic/_internal/_validate_call.py:
```diff
@@ -37,6 +37,11 @@ def __init__(self, function: Callable[..., Any], config: ConfigDict | None, vali
+        # specifically, we shouldn't be pumping the namespace full of type_params
+        # when we take namespace and type_params arguments in eval_type_backport
+        type_params = getattr(schema_type, '__type_params__', ())
+        namespace.update({param.__name__: param for param in type_params})
```
```suggestion
        namespace = {param.__name__: param for param in type_params} | namespace
```

This gives `type_params` lower precedence which is important

sydney-runkle replied:
Ah yeah, but the `|` syntax isn't supported in 3.8

sydney-runkle replied:
But I'll do dict unpacking to preserve that

sydney-runkle commented:
I'm unsure of why `pdm` is now having issues after I made two minor changes:

https://github.com/pydantic/pydantic/actions/runs/9668346229/job/26672294116?pr=9743

alexmojaki reviewed (changes requested):
I think this should be split into "support 3.13" and "support type params". The latter itself could be split for supporting different situations. In particular this doesn't cover parametrized classes at all, only `@validate_call`. Even for just `@validate_call`, this needs tests.

sydney-runkle commented:
@alexmojaki,

Sure, I'll pull out the `@validate_call` changes for the type param support.

sydney-runkle commented:
Edit: before, I had this in the description:

> Closes https://github.com/pydantic/pydantic/issues/9536
> 
> @alexmojaki, I've added your suggested changes for #9536 with an additional comment for the fix in `_validate_call.py` regarding the fact that we should probably change how we pass around those type params.
> 
> I think it's probably worth sprinting on providing full and clean support for PEP 695 stuff (though not all in this PR)

I've now moved this to a different PR :) (to be linked soon)

sydney-runkle commented:
I'd like to check in around the time of the official 3.13 release to remove some of the todos in `pyproject.toml` (testing packages that don't yet support 3.13, mostly).
